### PR TITLE
feat(wasm): credential signers for HMAC, EIP-712, NEP-413, Solana

### PR DESCRIPTION
**Pull request by neo-sky**
## Summary

Adds four new `CredentialLocation` variants (`HmacSignedHeader`, `Eip712SignedHeader`, `Nep413SignedHeader`, `SolanaSignedTransaction`) alongside the existing five static primitives. Tools declare venue-specific signing schemas in `capabilities.json` via a closed source vocabulary; the runtime stays generic. Unblocks Polymarket L1+L2, Hyperliquid, Trezu, NEAR Intents, and Solana RPC `sendTransaction` without per-venue runtime code.

## Motivation

The five existing `CredentialLocation` variants only support static placement of secrets into headers, query params, or URL placeholders. Real signing recipes need request data (method, path, body, timestamp, message bytes) that is only known at request time, plus crypto the WASM sandbox cannot safely run. This blocks every venue using HMAC, EIP-712 typed data, NEP-413, or Solana ed25519: Polymarket L1 (ClobAuth) + L2 (HMAC), Hyperliquid (EIP-712 with body-output signature), Trezu (NEP-413), NEAR Intents (NEP-413), Solana RPC (`sendTransaction` with a fully signed transaction in `params[0]`).

## Architecture

Four new variants on the existing `CredentialLocation` enum, same dispatcher loop (`inject_host_credentials`), same resolver (`resolve_host_credentials`), same schema layer. No architectural change.

Closed source vocabulary (`FieldSource`, `OutputSource`, `BodyValue`, `BytesPart`) bounds what tools can declare to be signed: every value is either user-reviewed at install time (manifest literal) or derived from the request the tool was already authorized to send (timestamp, nonce, body field, msgpack of body field, base64 message bytes). No "sign these arbitrary bytes" escape hatch exists.

Body mutation runs after the WASM tool hands over the request body but before reqwest dispatch, needed for Hyperliquid (which puts `signature.{r,s,v}` in the body) and Solana (which puts the full base64 signed transaction at `params.0` of the JSON-RPC envelope). The new `output_body_fields: Vec<BodyJsonOutput>` field declares JSON paths and value sources; the runtime parses the body once, walks the paths, sets the values, re-serializes. `json_path_set` now walks array indices so output can land at JSON-RPC params slots.

EIP-712 schemas in this PR are constrained to a single struct: only the primary type, no nested type dependencies. All four EVM target venues use single-struct schemas, so this is sufficient for the unblocking work. Multi-struct support per the full EIP-712 spec is a follow-up; the manifest format already accepts `Vec<Eip712StructDef>`, so adding nested-struct dependency resolution later is additive.

Solana signing follows the same closed-vocabulary pattern: the tool builds the unsigned wire-format message bytes itself (account ordering, recent_blockhash, instruction encoding) and places them base64-encoded in the request body via the new `FieldSource::BodyFieldString { path }` source. The runtime decodes, ed25519-signs the raw message bytes per RFC 8032, prepends the compact-u16 signature count and the 64-byte signature, base64-encodes, and writes the result via `output_body_fields` (typically into `params.0`). Total transaction size is capped at the Solana 1232-byte packet limit.

## Security

- All signing happens host-side. WASM tools never see HMAC keys, secp256k1 private keys, or ed25519 seeds.
- WIT contract unchanged: only `secret_exists(name)` is exposed to WASM, never `secret_get`.
- `LeakDetector` still scans WASM-provided URL/headers/body before any host injection.
- Sandbox proxy skip-list extended for the four signing variants alongside the existing `AuthorizationBasic`/`UrlPath` complex variants.
- `ResolvedHostCredential::Debug` redacts `secret_value` and shows only header NAMES (no values).
- Signing context (timestamp, random nonce) is generated once per request and shared across multiple matching signers.
- Solana signer rejects ECDSA r/s/v body outputs at runtime; the EIP-712 evaluator rejects `SolanaSignedTransactionBase64` outside the Solana signer. Cross-signer output type confusion fails fast.
- uint256 EIP-712 encoding uses a hand-rolled big-endian decimal parser supporting the full 2^256 - 1 range with overflow rejection.

## Tests

- 30 new tests covering all four signers, body mutation, schema deserialization, end-to-end pipeline through real capabilities manifests, uint256 encoding edge cases, and cross-signer output type confusion guards.
- Cryptographic round-trips:
  - HMAC: signature recomputed via the `hmac` crate and asserted equal.
  - EIP-712: signature recovered to the signer address via `VerifyingKey::recover_from_prehash` against the recomputed final hash (validates domain separator, type hash, struct hash).
  - NEP-413: signature verified via `ed25519_dalek::Verifier` over `sha256(prefix || borsh-payload)` (validates prefix bytes \`[0x9D, 0x01, 0x00, 0x80]\` matches \`2^31 + 413\`, borsh field order, hash).
  - Solana: signature verified via `ed25519_dalek::Verifier` directly against the raw message bytes (validates RFC 8032 application-level no-prehash, wire-format envelope assembly, base64 encoding).
- Polymarket-clob integration tests load the actual manifest content, resolve dummy secrets through `resolve_host_credentials`, and verify both L1 (\`/auth/api-key\` EIP-712) and L2 (\`/data/orders\` HMAC) paths assemble all expected headers correctly.
- Solana integration test loads a JSON-RPC sendTransaction body with the unsigned message at \`_signing.message_b64\` and verifies the assembled body has a valid signed transaction at \`params.0\` that ed25519-verifies against the signer pubkey.
- uint256 encoding regression test covers zero, value just above 2^128, the 2^256 - 1 maximum, the 2^256 overflow case, and non-numeric / negative / empty input rejection.
- Full lib suite: 5620 passed, 0 failed, 7 ignored.
- \`cargo clippy --lib --tests\`: zero warnings.

## Rollback

Additive only. New variants on existing enums, no DB migration, no schema changes, no breaking API changes. Tools that do not declare the new variants are unaffected.

## Spec references

- Polymarket: [py-clob-client signing/eip712](https://github.com/Polymarket/py-clob-client/blob/main/py_clob_client/signing/eip712.py), [py-clob-client headers](https://github.com/Polymarket/py-clob-client/blob/main/py_clob_client/headers/headers.py)
- NEP-413: [near/NEPs#413](https://github.com/near/NEPs/blob/master/neps/nep-0413.md)
- Hyperliquid: [hyperliquid-python-sdk signing.py](https://github.com/hyperliquid-dex/hyperliquid-python-sdk/blob/master/hyperliquid/utils/signing.py)
- Solana: [Transaction Structure](https://solana.com/docs/core/transactions/transaction-structure), [sendTransaction RPC](https://solana.com/docs/rpc/http/sendtransaction), [solana-short-vec compact-u16 source](https://docs.rs/solana-short-vec/latest/src/solana_short_vec/lib.rs.html)

### CODE REVIEW

**Review by zmanian — CHANGES_REQUESTED** [review]
# Review

**Summary:** Adds four new `CredentialLocation` variants letting WASM tools declare a closed-vocabulary signing schema in `capabilities.json`; the host performs all crypto with secrets that never cross the WASM boundary. Big PR (+3482/-28), well-architected.

## Crypto implementation is clean

- **EIP-712 domain is manifest-bound, not WASM-supplied.** `Eip712Domain { name, version, chain_id, verifying_contract }` is declared in `capabilities.json` as a `CredentialLocationSchema::Eip712SignedHeader` literal. The WASM tool cannot mutate `chainId` or `verifyingContract` at runtime — they're baked into the install-reviewed manifest. Trust anchor moves to manifest-review quality. **But only as strong as the install-time UX** (see below).
- **NEP-413 prefix correctly enforced host-side.** `apply_nep413_signing` hardcodes `prefix: u32 = 2_147_484_061` (= 2^31 + 413) and borsh-serializes it before the payload. Signing is `sha256(prefix || borsh(payload))`, ed25519 over the digest. WASM has no bypass path. Correct.
- **Solana key material never crosses WASM boundary.** `apply_solana_signing` runs host-side; secret resolution via `resolve_host_credentials` (host scope only). WIT exposes only `secret_exists`, never `secret_get`.
- **Input validation tight.** uint256 overflow rejected; Solana 1232-byte packet limit enforced; cross-signer body-value confusion rejected (`SolanaSignedTransactionBase64` outside Solana signer → `InvalidSchema`); secp256k1 v=27/28 convention preserved.
- **HMAC key isolation** relies on existing `SecretsStore` + `CredentialMapping.tool_id` scoping — not weakened.

## Architectural gaps — blocking for wallet-draining surfaces

1. **No per-request user consent.** Once a tool is installed with a `SolanaSignedTransaction` or `Eip712SignedHeader` capability, every subsequent invocation signs silently. The closed `FieldSource`/`BytesPart` vocabulary constrains the *shape* of the signed payload, not its *value* — a malicious/compromised extension can construct a Solana transaction draining the wallet within the allowed shape. The host signs raw `message_bytes` from `BodyFieldString` without parsing instructions.

   This is the largest unmitigated risk. Recommend a per-call user approval for high-value signers (Solana `sendTransaction`, EIP-712 OrderTyped) or at minimum a rate-limit + UI-surfaced audit log.

2. **Signing capability bundled with HTTP allowlist** at install time. No separate `sign:eip712` / `sign:solana` scope a user can revoke independently of the tool. Combined with (1), signing authority + network authority are bundled. A user can't say "I trust this for reading prices but not for signing trades."

## Verify separately

3. **Install-time EIP-712 domain rendering.** The whole #1 trust-anchor argument depends on `ironclaw extension install` surfacing the EIP-712 domain (chain, contract, version) to the user in **human-readable form** before approval. Not visible in this PR — confirm separately. Without it, "manifest-bound" reduces to "extension-author-bound."

4. **No host-side replay protection.** `request_random_nonce_b64` / `request_timestamp_secs` generated fresh per request but the host doesn't track used nonces. If a venue's protocol expects monotonic or one-shot nonces, enforcement is implicit in the venue API, not in IronClaw. Acceptable but document explicitly.

## Verdict

Crypto: ship. Architecture: not yet — the per-tool blank-cheque signer is a wallet-draining vector even with the closed vocabulary. Recommend opening tracking issues for (a) install-time EIP-712 domain rendering, (b) per-call approval for high-value signers, (c) splitting `sign:*` from HTTP allowlist as separate capability scopes, then re-reviewing. Given the blast radius, a human security pass before merge is also warranted.

**Comment by zmanian** [comment]
Filed follow-up security issue **#3564** capturing the architectural concern from my review:

> [Security] Wallet signing requires an unforgeable user-authorization channel, not host-resident keys
> https://github.com/nearai/ironclaw/issues/3564

The crypto in this PR is sound and useful for the lower-stakes paths (HMAC auth, NEP-413 message attestations, EIP-712 *off-chain* venue auth). The follow-up issue contests only the *wallet-equivalent* signing-authority use cases — where host compromise translates directly to unbounded fund loss — and proposes:

1. **Interactive sessions** → EIP-1193 / WalletConnect bridge (IronClaw composes, user's wallet signs).
2. **Autonomous missions** → session keys with on-chain caps (Permit2 / ERC-4337 / Solana PDA authorities). Drain blast radius ≤ delegation cap.
3. **Optional hardening** → TEE-resident signer + WebAuthn/passkey assertion for cap extension.
4. **Architectural rule** → no high-value chain-signing key lives in the IronClaw host process address space.

The manifest-bound EIP-712 domain stays a useful UX/ergonomics feature in that world; it just stops being a load-bearing security control.

Not a request to revert this PR — the in-scope crypto can land. The follow-up gates what these capabilities are *advertised* for and adds the missing channel for the value-bearing use cases.

File: src/tools/wasm/credential_injector.rs [hunk 1]
```diff
@@ -427,6 +427,10 @@ pub(crate) fn inject_credential(
             // URL placeholder replacement is handled by channel/tool wrappers
             // that substitute {PLACEHOLDER} values in templated strings.
         }
+        CredentialLocation::HmacSignedHeader { .. }
+        | CredentialLocation::Eip712SignedHeader { .. }
+        | CredentialLocation::Nep413SignedHeader { .. }
+        | CredentialLocation::SolanaSignedTransaction { .. } => {}
```
[the remark this answers is not in the document]
Comment (neo-sky) [reply src/tools/wasm/credential_injector.rs:533]: Intentional, documented inline in `309ac857`. Signing variants aren't injected here; signatures are computed host-side in the signer path after the body is assembled, so the secret never enters static header/query/url placement and the WASM tool never sees key material.

**Comment by neo-sky** [comment]
@zmanian thanks for the thorough review, that was genuinely useful. Pushed the changes (rebased on current staging, fmt/clippy/tests green):

Install-time EIP-712 rendering is in. The domain (name, version, chain in human-readable form, verifying contract) is shown and the user has to consent before approval, on both `ironclaw registry install` and `ironclaw tool install`. Known vs unknown domains get flagged.

Per-call approval for the high-value signers is wired end to end (app to ToolRegistry to the WASM wrappers). Solana is classified High, EIP-712 is graded by its struct shape. The gate fails closed when there's no TTY, so on a hosted or non-interactive agent a high-classification signer can't go through silently.

Signing is now its own capability section in capabilities.json, separate from the HTTP allowlist and with its own revoke path, so network access and signing authority aren't bundled together anymore.

On replay, I wrote the posture down explicitly at the nonce site rather than leaving it implied: fresh nonce and timestamp per request, the host does not track or enforce uniqueness, replay resistance is delegated to the venue API.

One thing worth calling out. While wiring the approval gate I found a real bug in the same area: when two signing schemas matched a request and wrote the same header it was silent last-writer-wins, so a low-classification HMAC signer was overwriting a higher-classification EIP-712 signature. It now resolves by classification precedence (higher wins, and the override is logged), and an equal-classification collision fails closed instead of silently picking one. Added regression tests, and a previously-failing end-to-end test now passes for the right reason.

The Gemini note about the empty match arm in `credential_injector` is answered: it's intentional, signing happens host-side after the body is assembled, and there's now an inline comment saying so.

**Review by zmanian — APPROVED** [review]
# Re-review (head `309ac85`, 2026-05-18) — follow-up to my 2026-05-13 CHANGES_REQUESTED

## What changed since my review (`d1df1dc` → `309ac85`)

Seven commits. The crypto layer I already cleared is unchanged; the new work targets exactly the architectural blockers I raised:

- `741c6d1f` **install consent + per-call approval gate + `sign:*` scope split** — the three blocker asks.
- `309ac857` **deterministic signer header-collision precedence** + documents the per-request replay posture and the intentionally-empty `credential_injector` signing arm.
- `2f8e38a8` / `583106fa` / `6f6c1ff9` — uint256 full-range parser, single body-JSON parse, shared signing context, Solana signer, polymarket-clob integration tests (refinements, not blockers).

## Status of my prior items

1. **Per-request user consent for high-value signers — RESOLVED.** `SignerApprovalGate` trait (`src/secrets/types.rs:366`) with a `classify_signing_location` classifier (`types.rs:290`): Solana → High, EIP-712 with address+value/order/permit/etc. → High, address-only → Medium, HMAC/NEP-413 → Low. The gate fires *before* any signing (`wrapper.rs:476`, ahead of the signer dispatch at `:478`), at a configurable floor defaulting to High (`wrapper.rs:2105`). `TerminalSignerGate` (`src/tools/wasm/signer_gate.rs`) prompts on a TTY and **fails closed** on non-interactive stdin (`signer_gate.rs:25-34`) and on read error. Production wiring confirmed end-to-end: `app.rs:537-539` installs `TerminalSignerGate` on the `ToolRegistry`, which propagates it to each WASM tool wrapper (`src/tools/registry.rs:999`, `wrapper.rs:2235`).
   - Note (non-blocking): `gate_signing_if_required` returns `Ok(())` when no gate is set (`wrapper.rs:566`). Safe today because the only production constructor wires a gate, but the *absence* of a gate is fail-open. Consider making High-classification signing require a gate by construction (type-level) so a future caller that forgets `with_signer_gate` can't silently sign. Tracking-issue-worthy, not merge-blocking.

2. **`sign:*` scope split from HTTP allowlist — RESOLVED.** Signing now lives on a dedicated `signing.schemes` axis (`capabilities_schema.rs:75`, `:336`), separate from `http.credentials`. Legacy signers smuggled under `http.credentials` are auto-promoted at parse time with a deprecation warning (`promote_legacy_signers`, `caps:136-159`). Users can reason about signing authority independently of network reach.

3. **Install-time EIP-712 domain rendering (was "verify separately") — RESOLVED.** `src/cli/signer_render.rs` renders domain (name/version/chainId with human chain names/`verifying_contract`), primary type, and every field's value source in human-readable form, classifies known vs unknown domains (`classify_domain`), and `prompt_for_signing_consent` requires the user to type an exact consent phrase (fails closed on non-TTY unless `--yes`). Wired at `src/cli/tool.rs:267-270`. Solana schemes render an explicit funds-transfer WARNING.

4. **Host-side replay protection (was "document explicitly") — RESOLVED as documented.** `309ac857` documents the per-request nonce/timestamp posture at the nonce site; enforcement remains venue-side, now explicitly stated. Acceptable.

## New findings

- **Security: none High/Critical.** Key material still never crosses the WASM boundary; `secret_get` is not in the WIT. The hand-rolled `Debug` redaction is intact with a guard comment against `#[derive(Debug)]` (`wrapper.rs:213-260`, `secret_value: "[REDACTED]"`), and `redact_credentials` scrubs secret values from logged request text (`wrapper.rs:356`). Signer error messages carry only error context, never secret bytes (`parse_secp256k1_secret`/`parse_ed25519_secret`, `wrapper.rs:995-1033`). Header-collision is now deterministic and **fails closed** on equal-classification ambiguity (`wrapper.rs:532` → refuses to sign).
- **Low / non-blocking — no zeroization of decoded key material.** `secret_value: String` and the decoded `bytes`/`seed_bytes` in the parse fns are not zeroized on drop. Consistent with the existing `SecretsStore` posture (plain `String`), so this PR introduces no regression; zeroizing only here would be incomplete. Worth a follow-up to thread `Zeroizing`/`SecretString` through the whole secret path.
- **Trivial.** `tracing::warn!` in `promote_legacy_signers`/collision paths — `warn!` is discouraged on REPL hot paths per repo convention, but these fire at tool-load/parse time and match the surrounding file's established use; collision warns are off the normal path. Not blocking.

## Tests
30+ new tests: per-signer crypto round-trips (HMAC recompute, EIP-712 address recovery, NEP-413/Solana ed25519 verify), uint256 edge cases, cross-signer output-type confusion guards, gate behavior (allow/reject/deny-on-non-TTY, tested through the wrapper at `wrapper.rs:7184+`), signer_render summaries, and polymarket-clob end-to-end through the real manifest. The gate is tested through the caller (the wrapper request path), not just the trait — satisfies test-through-the-caller. `classify`/`scope` CI checks green; PR mergeable.

## Verdict

**APPROVE.** All three blocking architectural items and both verify-separately items are resolved with the gate verified wired through to production and failing closed; the crypto layer I cleared previously is unchanged; no new High/Critical. The remaining items (fail-open-when-no-gate by construction, key-material zeroization) are genuine but non-blocking hardening — recommend tracking issues. Given the wallet-draining blast radius, a human security pass before merge is still prudent, but nothing here warrants holding the PR.

_Re-reviewed on behalf of @zmanian._

**Comment by serrrfirat** [comment]
Thank you for the work and investigation in this PR. We are closing it as part of the transition from the legacy IronClaw runtime to the Reborn architecture.

The credential signer variants and injection logic are implemented in the legacy WASM capability/secret path. Reborn signing and credential use are governed by typed authorization obligations, runtime credential audiences, approval gates, and host-mediated egress; forwarding guest-controlled signer injection would bypass those boundaries.

The signer formats can still contribute. Please add individual signers behind the current authorization/signing ports, with opaque host-held keys, explicit network/transaction approval, domain-separated payloads, and caller-level tests. Coordinate with the active Reborn attested-signing stack rather than creating a parallel credential vocabulary.

We would be glad to review a focused Reborn contribution along those lines. Please start from the current crate contracts and production composition rather than rebasing the legacy implementation mechanically. Thank you again for contributing and for helping expose the underlying product need.

**Comment by serrrfirat** [comment]
Thank you for the work in this PR. We are closing it as part of the transition from the legacy IronClaw runtime to Reborn.

These signer and credential-injection changes use the legacy WASM secret path. Reborn uses authorization obligations, runtime credential audiences, approval gates, and host-mediated egress.

Please contribute signer formats through the current Reborn signing/authorization ports, using opaque host keys, domain-separated payloads, explicit transaction approval, and caller-level tests.

We would be glad to review a focused Reborn contribution built on the current crate contracts and production composition. Thank you again for contributing and for identifying the underlying need.

**Comment by serrrfirat** [comment]
Thank you for the work in this PR. We are closing it as part of the transition from the legacy IronClaw runtime to Reborn.

These signer and credential changes use the legacy WASM secret path. Reborn uses authorization obligations, credential audiences, approval gates, and host-mediated egress.

Please contribute signer formats through current Reborn signing ports with opaque keys, domain separation, transaction approval, and caller-level tests.

We would be glad to review a focused Reborn contribution built on current crate contracts. Thank you again for contributing.
